src/main/java/ruslan/shastkiv/bookstore/service/cart/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,87 @@
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemRepository cartItemRepository;
+    private final BookService bookService;
+
+    @Override
+    public void createShoppingCart(User user) {
+        ShoppingCart cart = new ShoppingCart();
+        cart.setUser(user);
+        shoppingCartRepository.save(cart);
+    }
+
+    @Override
+    public ShoppingCartDto getShoppingCart(Long userId) {
+        return shoppingCartMapper.toDto(findShoppingCart(userId));
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartDto addBookToCart(Long userId, CartItemRequestDto requestDto) {
+        ShoppingCart shoppingCart = findShoppingCart(userId);
+        return shoppingCartMapper.toDto(addCartItem(shoppingCart, requestDto));
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartDto updateItemQuantity(Long userId,
+                                               Long cartItemId,
+                                               UpdateCartItemRequestDto requestDto) {
+        CartItem cartItem = getByCartAndUserId(userId, cartItemId);
+        cartItem.setQuantity(requestDto.quantity());
+        return shoppingCartMapper.toDto(
+                shoppingCartRepository.save(cartItem.getShoppingCart()));
+    }
+
+    @Override
+    public void removeCartItem(Long userId, Long cartItemId) {
+        CartItem cartItem = getByCartAndUserId(userId, cartItemId);
+        cartItemRepository.delete(cartItem);
+    }
+
+    private ShoppingCart findShoppingCart(Long id) {
+        return shoppingCartRepository.findById(id).orElseThrow(()
+                -> new EntityNotFoundException("Cant find shopping cart by user id: " + id));
+    }
+
+    private CartItem getByCartAndUserId(Long userId, Long cartItemId) {
+        return cartItemRepository.findByIdAndShoppingCartId(cartItemId, userId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Cart item with ID [ " + cartItemId
+                                + " ] or Shopping Cart for user ID [ " + userId + " ] not found"));
+    }
+
+    private ShoppingCart addCartItem(ShoppingCart shoppingCart, CartItemRequestDto requestDto) {
+        CartItem cartItem = cartItemRepository
+                .findByBookIdAndShoppingCartId(requestDto.bookId(), shoppingCart.getId())
+                .orElseGet(() -> {
+                    CartItem newCartItem = new CartItem();
+                    newCartItem.setBook(bookService.findBookById(requestDto.bookId()));
+                    newCartItem.setShoppingCart(shoppingCart);
+                    shoppingCart.getCartItems().add(newCartItem);
+                    return newCartItem;
+                });
+        cartItem.setQuantity(cartItem.getQuantity() + requestDto.quantity());
+        return shoppingCartRepository.save(shoppingCart);
```

Comment:
I think you should not update quantity, because It won`t be possible to decrease amount of requested book in cart